File: _edgedb/queries.py
#
# Copyright (c) 2019 MagicStack Inc.
# All rights reserved.
#
# See LICENSE for details.
##


import edgedb
import random


INSERT_PREFIX = 'insert_test__'


def connect(ctx):
    return edgedb.connect()


def close(ctx, conn):
    conn.close()


def load_ids(ctx, conn):
    d = conn.query_single('''
        WITH
            U := User {id, r := random()},
            M := Movie {id, r := random()},
            P := Person {id, r := random()}
        SELECT (
            users := array_agg((SELECT U ORDER BY U.r LIMIT <int64>$lim).id),
            movies := array_agg((SELECT M ORDER BY M.r LIMIT <int64>$lim).id),
            people := array_agg((SELECT P ORDER BY P.r LIMIT <int64>$lim).id),
        );
    ''', lim=ctx.number_of_ids)

    return dict(
        get_user=list(d.users),
        get_movie=list(d.movies),
        get_person=list(d.people),
        # re-use user IDs for update tests
        update_movie=list(d.movies),
        # generate as many insert stubs as "concurrency" to
        # accommodate concurrent inserts
        insert_user=[INSERT_PREFIX] * ctx.concurrency,
    )


def get_user(conn, id):
    return conn.query_single_json('''
        SELECT User {
            id,
            name,
            image,
            latest_reviews := (
                WITH UserReviews := User.<author[IS Review]
                SELECT UserReviews {
                    id,
                    body,
                    rating,
                    movie: {
                        id,
                        image,
                        title,
                        avg_rating
                    }
                }
                ORDER BY .creation_time DESC
                LIMIT 10
            )
        }
        FILTER .id = <uuid>$id
    ''', id=id)


def get_movie(conn, id):
    return conn.query_single_json('''
        SELECT Movie {
            id,
            image,
            title,
            year,
            description,
            avg_rating,

            directors: {
                id,
                full_name,
                image,
            }
            ORDER BY Movie.directors@list_order EMPTY LAST
                THEN Movie.directors.last_name,

            cast: {
                id,
                full_name,
                image,
            }
            ORDER BY Movie.cast@list_order EMPTY LAST
                THEN Movie.cast.last_name,

            reviews := (
                SELECT Movie.<movie[IS Review] {
                    id,
                    body,
                    rating,
                    author: {
                        id,
                        name,
                        image,
                    }
                }
                ORDER BY .creation_time DESC
            ),
        }
        FILTER .id = <uuid>$id
    ''', id=id)


def get_person(conn, id):
    return conn.query_single_json('''
        SELECT Person {
            id,
            full_name,
            image,
            bio,

            acted_in := (
                WITH M := Person.<cast[IS Movie]
                SELECT M {
                    id,
                    image,
                    title,
                    year,
                    avg_rating
                }
                ORDER BY .year ASC THEN .title ASC
            ),

            directed := (
                WITH M := Person.<directors[IS Movie]
                SELECT M {
                    id,
                    image,
                    title,
                    year,
                    avg_rating
                }
                ORDER BY .year ASC THEN .title ASC
            ),
        }
        FILTER .id = <uuid>$id
    ''', id=id)


def update_movie(conn, id):
    return conn.query_single_json('''
        SELECT (
            UPDATE Movie
            FILTER .id = <uuid>$id
            SET {
                title := .title ++ '---' ++ <str>$suffix
            }
        ) {
            id,
            title
        }
    ''', id=id, suffix=str(id)[:8])


def insert_user(conn, val):
    num = random.randrange(1_000_000)
    return conn.query_single_json('''
        SELECT (
            INSERT User {
                name := <str>$name,
                image := <str>$image,
            }
        ) {
            id,
            name,
            image,
        }
    ''', name=f'{val}{num}', image=f'image_{val}{num}')


def setup(ctx, conn, queryname):
    if queryname == 'update_movie':
        conn.execute('''
            update Movie
            filter contains(.title, '---')
            set {
                title := str_split(.title, '---')[0]
            };
        ''')
    elif queryname == 'insert_user':
        conn.query('''
            delete User
            filter .name LIKE <str>$prefix
        ''', prefix=f'{INSERT_PREFIX}%')


def cleanup(ctx, conn, queryname):
    if queryname in {'update_movie', 'insert_user'}:
        # The clean up is the same as setup for mutation benchmarks
        setup(ctx, conn, queryname)

File: _elixir/lib/bench/postgrex/queries.ex
defmodule Bench.Postgrex.Queries do
  @get_user """
  SELECT
      users.id,
      users.name,
      users.image,
      q.review_id,
      q.review_body,
      q.review_rating,
      q.movie_id,
      q.movie_image,
      q.movie_title,
      q.movie_avg_rating
  FROM
      users,
      LATERAL (
          SELECT
              review.id AS review_id,
              review.body AS review_body,
              review.rating AS review_rating,
              movie.id AS movie_id,
              movie.image AS movie_image,
              movie.title AS movie_title,
              movie.avg_rating AS movie_avg_rating
          FROM
              reviews AS review
              INNER JOIN movies AS movie
                  ON (review.movie_id = movie.id)
          WHERE
              review.author_id = users.id
          ORDER BY
              review.creation_time DESC
          LIMIT 10
      ) AS q
  WHERE
      users.id = $1
  """

  @get_person """
  SELECT
      person.id,
      person.full_name,
      person.image,
      person.bio,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
        FROM
          (SELECT
              ROW(
                  movie.id,
                  movie.image,
                  movie.title,
                  movie.year,
                  movie.avg_rating
              ) AS v
          FROM
              actors
              INNER JOIN movies AS movie
                  ON (actors.movie_id = movie.id)
          WHERE
              actors.person_id = person.id
          ORDER BY
              movie.year ASC, movie.title ASC
          ) AS q
      ) AS acted_in,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
        FROM
          (SELECT
              ROW(
                  movie.id,
                  movie.image,
                  movie.title,
                  movie.year,
                  movie.avg_rating
              ) AS v
          FROM
              directors
              INNER JOIN movies AS movie
                  ON (directors.movie_id = movie.id)
          WHERE
              directors.person_id = person.id
          ORDER BY
              movie.year ASC, movie.title ASC
          ) AS q
      ) AS directed
  FROM
      persons AS person
  WHERE
      id = $1;
  """

  @get_movie """
  SELECT
      movie.id,
      movie.image,
      movie.title,
      movie.year,
      movie.description,
      movie.avg_rating,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
        FROM
          (SELECT
              ROW(
                  person.id,
                  person.full_name,
                  person.image
              ) AS v
          FROM
              directors
              INNER JOIN persons AS person
                  ON (directors.person_id = person.id)
          WHERE
              directors.movie_id = movie.id
          ORDER BY
              directors.list_order NULLS LAST,
              person.last_name
          ) AS q
      ) AS directors,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
        FROM
          (SELECT
              ROW(
                  person.id,
                  person.full_name,
                  person.image
              ) AS v
          FROM
              actors
              INNER JOIN persons AS person
                  ON (actors.person_id = person.id)
          WHERE
              actors.movie_id = movie.id
          ORDER BY
              actors.list_order NULLS LAST,
              person.last_name
          ) AS q
      ) AS actors,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
        FROM
          (SELECT
              ROW(
                  review.id,
                  review.body,
                  review.rating,
                  (SELECT
                      ROW(
                          author.id,
                          author.name,
                          author.image
                      )
                      FROM
                          users AS author
                      WHERE
                          review.author_id = author.id
                  )
              ) AS v
          FROM
              reviews AS review
          WHERE
              review.movie_id = movie.id
          ORDER BY
              review.creation_time DESC
          ) AS q
      ) AS reviews
  FROM
      movies AS movie
  WHERE
      id = $1;
  """

  @update_movie """
  UPDATE
    movies
  SET
    title = movies.title || $2
  WHERE
    movies.id = $1
  RETURNING
    movies.id, movies.title
  """

  @insert_user """
  INSERT INTO users (name, image) VALUES
    ($1, $2)
  RETURNING
    users.id, users.name, users.image
  """

  @insert_movie """
  WITH
  m AS (
      INSERT INTO movies AS M (title, image, description, year)
      VALUES ($1, $2, $3, $4)
      RETURNING M.id, M.title, M.image, M.description, M.year
  ),
  d AS (
      SELECT
          person.id,
          person.full_name,
          person.image
      FROM
          persons AS person
      WHERE
          id = $5
  ),
  c AS (
      SELECT
          person.id,
          person.full_name,
          person.image
      FROM
          persons AS person
      WHERE
          id IN ($6, $7, $8)
  ),
  dl AS (
      INSERT INTO directors (person_id, movie_id)
      (SELECT d.id, m.id FROM m, d)
  ),
  cl AS (
      INSERT INTO actors (person_id, movie_id)
      (SELECT c.id, m.id FROM m, c)
  )
  SELECT
      m.id,
      m.image,
      m.title,
      m.year,
      m.description,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
      FROM
          (SELECT
              ROW(id, full_name, image) AS v
          FROM
              d
          ) AS q
      ) AS directors,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
      FROM
          (SELECT
              ROW(id, full_name, image) AS v
          FROM
              c
          ) AS q
      ) AS actors
  FROM
      m
  """

  @insert_movie_plus """
  WITH
  m AS (
      INSERT INTO movies AS M (title, image, description, year)
      VALUES ($1, $2, $3, $4)
      RETURNING M.id, M.title, M.image, M.description, M.year
  ),
  p AS (
      INSERT INTO persons AS P (first_name, last_name, image, bio)
      VALUES
          ($5, $6, $7, ''),
          ($8, $9, $10, ''),
          ($11, $12, $13, '')
      RETURNING
          P.id, P.last_name, P.full_name, P.image
  ),
  dl AS (
      INSERT INTO directors (person_id, movie_id)
      (SELECT p.id, m.id FROM m, p WHERE p.last_name = 'Director')
  ),
  cl AS (
      INSERT INTO actors (person_id, movie_id)
      (SELECT p.id, m.id FROM m, p WHERE p.last_name = 'Actor')
  )
  SELECT
      m.id,
      m.image,
      m.title,
      m.year,
      m.description,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
      FROM
          (SELECT
              ROW(id, full_name, image) AS v
          FROM
              p
          WHERE
              p.last_name = 'Director'
          ) AS q
      ) AS directors,
      (SELECT
          COALESCE(array_agg(q.v), (ARRAY[])::record[])
      FROM
          (SELECT
              ROW(id, full_name, image) AS v
          FROM
              p
          WHERE
              p.last_name = 'Actor'
          ) AS q
      ) AS actors
  FROM
      m
  """

  def get_user, do: @get_user
  def get_person, do: @get_person
  def get_movie, do: @get_movie
  def update_movie, do: @update_movie
  def insert_user, do: @insert_user
  def insert_movie, do: @insert_movie
  def insert_movie_plus, do: @insert_movie_plus
end
